### ISSUE TITLE
Unable to set logical sector size = 2048 bytes

### ISSUE THREAD

**Issue by kefir-** [karelzak/util-linux#396]
I have a dump of a raw memory card from a 32GB Sony video camera (SxS card). The dump has a DOS/MBR type partition, and a FAT32 file system in the first and only partition.

The issue now is that the FAT32 file system has a logical sector size of 2048 bytes. Here's a few commands showing the structure:

```
$ file - < <(dd if=image.dd bs=512 count=2)
/dev/stdin: DOS/MBR boot sector; partition 1 : ID=0xc, start-CHS (0x0,1,1), end-CHS (0x3cd,245,40), startsector 16, 15646704 sectors

$ file - < <(dd if=image.dd bs=512 count=2 skip=64)
/dev/stdin: DOS/MBR boot sector, code offset 0x3e+2, OEM-ID "MSWIN4.1", Bytes/sector 2048, sectors/cluster 16, Media descriptor 0xf8, sectors/track 63, heads 255, hidden sectors 16, sectors 15646704 (volumes > 32 MB) , FAT (32 bit), sectors/FAT 8144, serial number 0x0, label: "           "

$ fdisk -l image.dd
Disk image.dd: 29,9 GiB, 32044482560 bytes, 62586880 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xffffffff

Device        Boot start      End  Sectors  Size Id Type
image.dd              16 15646719 15646704  7,5G  c W95 FAT32 (LBA)
```
When I use losetup and partx to set this file up as a block device with partitions, the logical sector size is not taken into account, and the partition is intepreted as being only around 8GB, instead of 32GB. I don't see any information in the partition about the logical sector size, only in the file system itself, but the same card worked fine on Windows, in the camera, and with other tools I've tested to interpret the file system from the raw copy of the memory card.

Interestingly, if I run losetup - o $offset to configure the single partition directly, it mounts fine:
```
$ sudo losetup -r -o $(( 512 * 64 )) /dev/loop0 image.dd
$ sudo mount -o ro /dev/loop0 /mnt/tmp
$ df -h /mnt/tmp
Filesystem     Size   Used   Avail  Use%   Mounted on
/dev/loop0      30G    13G     18G   43%   /mnt/tmp
```
But having to resort to this method takes away the benfits of using losetup/partx to discover all the details, and takes be back to manually parsing all the partition and file system details.

Is there any way to override the logical sector size? It seems sensible that this should be an argument to losetup.

FWIW, I experimented a bit trying to get mdadm to create a proxy device with a different block size. Unfortunately it doesn't allow a smaller chunk size than 4k, but even when I tried setting up a linear device with only one component device (loop0 in my case) it turns out that the physical/logical sector size are both still 512 bytes.

**Comment by karelzak** [comment]
Good catch, it helps me found many fdisk bugs ;-)

Unfortunately, system reads sector size from block device HW rather than from content on the device. If you use only disk image than the information about sector size is missing.

The fdisk command supports "--sector-size" option, but it seems broken (fixed in git tree a few minutes ago). This option does not help you because it's used for PT parsing.

We need --sector-size for partx. I'll try to implement it. Thanks!

**Comment by kefir-** [comment]
Thanks for the update! I ran fdisk directly on the file image above, but I could run it against /dev/loop0 with the same result. Ideally I would like to be able to set up `/dev/loop0` to point to `image.dd`, then use partx to discover all the partitions and mount them from there. In this case there was just one partition, but that's not always the case.

I see how it makes sense to specify sector size to fdisk in the example given, but I also think it would make sense to specify sector size as an argument to losetup. Then a loop block device pointing to a flat file would appear to have the same physical parameters as the original block device the disk image comes from. I would think that changing this in losetup would cause everything else to "just work" after that.

**Comment by karelzak** [comment]
OK, with the new partx,  /dev/sdc is 4K disk:
 ``` 
     # fdisk -l /dev/sdc
     Disk /dev/sdc: 100 MiB, 104857600 bytes, 25600 sectors
     ...
     Device     Boot Start   End Sectors Size Id Type
     /dev/sdc1        1024 25599   24576  96M 83 Linux
```    
    let's create disk image:
```    
     # dd if=/dev/sdc of=~/sdc.img
     # losetup -f ~/sdc.img
```
Note that kernel may automatically create partition /dev/loop0p1, use lsblk to see if the partition has proper size -- probably it will be incorrect, because kernel has no clue about the proper sector size (no HW, just image). Use:
```
    partx --delete --nr 1 /dev/loop0
```
to remove this incorrect partition (if necessary), and use proper sector size:
```
   partx --add /dev/loop0 --sector-size 4096
```
and see lsblk:
```
NAME      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
loop0       7:0    0  100M  0 loop 
└─loop0p1 259:1    0   96M  0 loop 
```
mount the device...

**Comment by karelzak** [comment]
Unfortunately, it's not possible to set sector size for losetup, this is unsupported by kernel.

**Comment by kefir-** [comment]
Thanks, I look forward to trying out the new feature! I'll let you know!

**Comment by kefir-** [comment]
Specifying "-S 2048" worked like a charm with my disk image! Thanks. Now I just have to find a way to reliably discover the sector size from a dd image.

By the way, you had a 100MB device, /dev/sdc in your example, with a 4K sector size. That seems like some sort of emulated device. I'd be interested to hear how you set that up.

**Comment by karelzak** [comment]
Yes, I use scsi_debug kernel module. Unfortunately, it allows you to create only one disk and all is stored in RAM, but good enough for testing.
```
# Create "desktop-class" 4K drive
# (logical_block_size=512, physical_block_size=4096, alignment_offset=0):
modprobe scsi_debug dev_size_mb=$DEV_SIZE sector_size=512 physblk_exp=3
         
# Create "desktop-class" 4K drive w/ 63-sector DOS partition compensation
# (logical_block_size=512, physical_block_size=4096, alignment_offset=3584):
modprobe scsi_debug dev_size_mb=$DEV_SIZE sector_size=512 physblk_exp=3 lowest_aligned=7
      
# Create "enterprise-class" 4K drive
# (logical_block_size=4096, physical_block_size=4096, alignment_offset=0):
modprobe scsi_debug dev_size_mb=$DEV_SIZE sector_size=4096
```

and "rmmod scsi_debug" removed the disk. You can create partitions, filesystem, mount, whatever. 

For partition tables created by standard Linux tools you can expect 1MiB offset for the begin of the first partition -- it means 2048 sectors for 512bytes or 256 sectors for 4K disks. Maybe you can use the same heuristic for another disks too.

**Comment by kefir-** [comment]
Thanks for the tips and suggestions, much appreciated!

**Comment by kefir-** [comment]
By the way, perhaps iSCSI could be a way around this as well, to test multiple sector sizes. It might be a bit cumbersome, but there's a parameter to set the blocksize. From `man tgtadm`:
```
       -y, --blocksize <size>
           Block devices are created with a default block size of 512 bytes.
           This argument can be used to create block devices with different
           block sizes.

           Example:

           tgtadm --lld iscsi --mode logicalunit --op new \
                  --tid <TID> --lun <LUN> \
                  -b <backing-file> --blocksize=4096
```

**Comment by kefir-** [comment]
Actually, this seems to work quite nicely. I'm on Ubuntu, and I've installed `tgt` and `open-iscsi`, and done something like the following to set up the same disk image as above. The file image became `/dev/sde`:
```
$ cat /etc/tgt/conf.d/myimage.conf
<target iqn.2017-01.com.example:myimage>
  <backing-store /path/to/image.dd>
    block-size 2048
  </backing-store>
</target>

$ sudo service tgt reload
$ sudo iscsiadm -m discovery -t st -p localhost
127.0.0.1:3260,1 iqn.2017-01.com.example:myimage
$ sudo iscsiadm -m node --login
Logging in to [iface: default, target: iqn.2017-01.com.example:myimage, portal: 127,.0.0.1,3260] (multiple)
Login to [iface: default, target: iqn.2017-01.com.example:myimage, portal: 127,.0.0.1,3260] successful.
$ sudo fdisk -l /dev/sde
Disk /dev/sde: 29,9 GiB, 32044482560 bytes, 15646720 sectors
Units: sectors of 1 * 2048 = 2048 bytes
Sector size (logical/physical): 2048 bytes / 4096 bytes
I/O size (minimum/optimal): 4096 bytes / 4096 bytes
Disklabel type: dos
Disk identifier: 0xffffffff

Device        Boot start      End  Sectors   Size Id Type
/dev/sde1             16 15646719 15646704  29,9G  c W95 FAT32 (LBA)
```

It seems the physical sector size is always 4096 in this configuration, but this doesn't seem to matter for what I'm doing.

**Comment by pali** [comment]
> Unfortunately, it's not possible to set sector size for losetup, this is unsupported by kernel.

Is there a feature request for this?

**Comment by karelzak** [comment]
See commit a1a41597bfd55e709024bd91aaf024159362679c. It's already implemented in by kernel and supported by losetup (v2.31).

**Comment by anon99** [comment]
ioctl BLKSZGET was added, losetup gets --sector-size 4096 (etc), partx also has --sector-size. So losetup --sector-size 2048 --show -f -somefile-, then partx (may need --sector-size) and fdisk/gdisk /dev/loopx should report sector sizes correctly and parse the partition table.
